src/upgrade/client_request.rs

```diff
@@ -110,11 +110,36 @@ impl Decoder for Codec {
 
         let request_len = status.unwrap();
 
-        let ws_accept = ClientRequest::parse(|name| {
-            let h = headers.iter().find(|h| h.name.eq_ignore_ascii_case(name))?;
-            parse_str(h.value).ok()
-        })?
-        .ws_accept();
+        let mut builder = http::request::Builder::new();
+        if let Some(m) = request.method {
+            let method = http::method::Method::from_bytes(m.as_bytes())
+                .map_err(|e| Error::InvalidRequest(e.into()))?;
+            builder = builder.method(method);
+        }
+        if let Some(uri) = request.path {
+            builder = builder.uri(uri);
+        }
+        match request.version {
+            Some(0) => builder = builder.version(http::Version::HTTP_10),
+            Some(1) => builder = builder.version(http::Version::HTTP_11),
+            _ => Err(Error::Parsing(httparse::Error::Version))?,
+        }
+        for h in headers {
+            if h.name.is_empty() {
+                continue;
+            }
```

Comment:
Just use `http::HeaderName::from_str` and do it analogous to `server_response.rs` (i.e. pre-allocate the header map). `http` will return an error if the header length is 0, and so should this instead of ignoring it.

Reply:
Because of the way `httparse` works there will always be empty headers except if there are more than 64 headers. So I don't think it is a good idea to return that error.

Reply:
This also means that the incoming request will always allocate 25 fields even if there is less requests. It is also odd that 64 headers is allowed on outgoing requests, but only 25 on incoming so maybe that should also be changed.

Reply:
> Because of the way `httparse` works there will always be empty headers except if there are more than 64 headers. So I don't think it is a good idea to return that error.
> 
> ```rust
> pub const EMPTY_HEADER: Header<'static> = Header { name: "", value: b"" };
> ```

Another look might show that I am wrong on this and it was because I used the headers and not the request.headers which have been shrunk.